src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapelessCraftingRecipe.java

```diff
@@ -0,0 +1,160 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.item.recipe.crafting;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.item.crafting.ShapelessRecipes;
+import net.minecraft.util.NonNullList;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.inventory.Slot;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.item.recipe.crafting.ShapelessCraftingRecipe;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.SpongeImplHooks;
+import org.spongepowered.common.item.inventory.util.ItemStackUtil;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class SpongeShapelessCraftingRecipe extends ShapelessRecipes implements ShapelessCraftingRecipe {
+
+    private final ItemStackSnapshot exemplaryResult;
+    private final List<Predicate<ItemStackSnapshot>> ingredients;
+
+    public SpongeShapelessCraftingRecipe(ItemStackSnapshot exemplaryResult, List<Predicate<ItemStackSnapshot>> ingredients) {
+        super(ItemStackUtil.fromSnapshotToNative(exemplaryResult), null);
+
+        checkNotNull(exemplaryResult, "exemplaryResult");
+        checkArgument(exemplaryResult != ItemStackSnapshot.NONE, "exemplaryResult");
+
+        ingredients.forEach(ingredient ->
+                checkNotNull(ingredient, "The ingredient list must not contain null values."));
+
+        this.exemplaryResult = exemplaryResult;
+        this.ingredients = ImmutableList.copyOf(ingredients);
+    }
+
+    @Override
+    public ItemStackSnapshot getExemplaryResult() {
+        return this.exemplaryResult;
+    }
+
+    @Override
+    public List<Predicate<ItemStackSnapshot>> getIngredientPredicates() {
+        return this.ingredients;
+    }
+
+    @Override
+    public boolean isValid(GridInventory grid, World world) {
+        List<Predicate<ItemStackSnapshot>> ingredients = Lists.newArrayList(this.ingredients);
+
+        for (int y = 0; y < grid.getRows(); y++) {
+            byIncreasingTheSlotIndex:
+            for (int x = 0; x < grid.getColumns(); x++) {
+                ItemStackSnapshot itemStackSnapshot = grid.getSlot(x, y)
+                        .flatMap(Slot::peek).map(ItemStack::createSnapshot)
+                        .orElse(ItemStackSnapshot.NONE);
+
+                if (itemStackSnapshot == ItemStackSnapshot.NONE) {
+                    continue;
+                }
+
+                Iterator<Predicate<ItemStackSnapshot>> iterator = ingredients.iterator();
```

Comment:
This logic is lacking comments to see what is actually going on (Yes, it's valid checking, but how is it valid checking?)